flink-tests/src/test/java/org/apache/flink/test/checkpointing/ApproximateLocalRecoveryDownstreamITCase.java

```diff
@@ -0,0 +1,274 @@
+			this.maxParallelism = maxParallelism;
+			this.numberOfChannels = numberOfChannels;
+
+			String shortString = "I am a very long string to test partial records hohoho hahaha ";
+			StringBuilder builder = new StringBuilder(shortString);
+
+			for (int i = 0; i <= 2 * bufferSize / shortString.length() + 1; i++) {
+				builder.append(shortString);
+			}
+			this.longOrShortString = builder.toString();
+		}
+
+		@Override
+		public void run(SourceContext<Tuple4<Integer, Long, Integer, String>> ctx) throws Exception{
+			long timestamp = 1593575900000L;
+			while (running) {
+				synchronized (ctx.getCheckpointLock()) {
+					if (index % 100 == 0) {
+						Thread.sleep(500);
+					}
+					ctx.collect(new Tuple4<>(index, timestamp++, channelIndex(index), longOrShortString));
+				}
+				index++;
+			}
+		}
+
+		@Override
+		public void cancel() {
+			running = false;
+		}
+
+		private int channelIndex(int key) {
+			return KeyGroupRangeAssignment.assignKeyToParallelOperator(key, maxParallelism, numberOfChannels);
+		}
+	}
+
+	private static class FailingMapper<T> extends RichMapFunction<T, T>  {
+		private static final long serialVersionUID = 6334389850158703L;
+
+		public static volatile boolean failedBefore;
+
+		private final int failCount;
+		private int numElementsTotal;
+
+		private boolean failer;
+
+		public FailingMapper(int failCount) {
+			this.failCount = failCount;
+		}
+
+		@Override
+		public void open(Configuration parameters) {
+			failer = getRuntimeContext().getIndexOfThisSubtask() == 0;
+		}
+
+		@Override
+		public T map(T value) throws Exception {
+			numElementsTotal++;
+
+			if (!failedBefore) {
+				Thread.sleep(10);
+
+				if (failer && numElementsTotal >= failCount) {
+					failedBefore = true;
+					throw new Exception("Artificial Test Failure");
+				}
+			}
+
+			return value;
+		}
+	}
+
+	private static class ValidatingAtMostOnceSink extends RichSinkFunction<Tuple4<Integer, Long, Integer, String>> {
+
+		private static final long serialVersionUID = 1748426382527469932L;
+		private final int numElementsTotal;
+		private final BitSet duplicateChecker = new BitSet();
+		private final int[] numElements;
+		private final int numberOfInputChannels;
+
+		public ValidatingAtMostOnceSink(int numElementsTotal, int numberOfInputChannels) {
+			this.numElementsTotal = numElementsTotal;
+			this.numberOfInputChannels = numberOfInputChannels;
+			this.numElements = new int[numberOfInputChannels];
+		}
+
+		public ValidatingAtMostOnceSink(int numElementsTotal) {
+			this.numElementsTotal = numElementsTotal;
+			this.numberOfInputChannels = 1;
+			this.numElements = new int[numberOfInputChannels];
+		}
+
+		@Override
+		public void invoke(Tuple4<Integer, Long, Integer, String> value) throws Exception {
+			assert value.f2 < numberOfInputChannels;
+			numElements[value.f2]++;
+
+			if (duplicateChecker.get(value.f0)) {
+				throw new Exception("Received a duplicate: " + value);
+			}
```

Comment:
How will the `duplicateChecker` work if the task gets restarted? Won't this effectively mean that we are checking that we don't send duplicate message during normal operation?

Reply:
I was originally thinking to make it check non-duplication/non-loss after reconnection.

But you are right, it just checks normal processing is correct, no need to do it.